### ISSUE TITLE
Refactoring / Re-Design

### ISSUE THREAD

**Issue by Bonifatius94** [EnricoReg/robot-sf#1]
Hi @EnricoReg and @matteocaruso1993,

I've come across your project because I'm working on a very similar topic for my master thesis at the University of Augsburg where I'm trying to teach a robot to avoid pedestrians on a sidewalk while delivering e.g. a parcel to some client.

Your project looks really promising so far because you've already put all those tools together that are necessary to start training with OpenAI stable baselines, etc. The code is just a bit messy and slow, but I'll definitely figure that out. So be prepared that I'll be proposing the results of a major rewrite as a PR soon, according to your GPL license.

Please let me know if you're interested in any kind of cooperation. And it would be nice if I could cite you in my thesis as well (I didn't find anything on the internet).

Best wishes,
Marco

**Comment by EnricoReg** [comment]
Hi Marco,

thanks for contacting us. I'm glad you found our work useful, and I personally welcome the opportunity to have some kind of collaboration.
Due to professional and personal reasons, Matteo and I were forced to put aside the project one year ago. However, in the last couple of weeks we have decided to get back on the project to finalize a publication which was almost ready for submission at the time.

I agree the code is a bit messy, as we didn't have much Python and OOP experience back then. It was meant to work with the asynchronous RL infrustructure [asynch-rl](https://github.com/EnricoReg/asynch-rl) 
If you are able to improve the robot_sf environment code (and computational speed), it could be beneficial also from our side in order to re-train the RL model.

KR
Enrico

**Comment by Bonifatius94** [comment]
Hi @EnricoReg, that would be great if we can collaborate.

I've already reworked great parts of robot.py and robot_env.py. Now, I'm looking to improve the raycast logic inside the BinaryOccupancyGrid and the RangeSensor. Once that's done, the PySocialForce simulator extension will be my next target. And setting up a profiler and unit tests would be nice as well to improve code quality / performance.

Overall, I did remove lots of dead code (e.g. error cases that never happen) such that the program becomes easier to reason about and added pytest / pylint automation. Let's see where it goes.

Best wishes, Marco

**Comment by matteocaruso1993** [comment]
Hi Marco,

Thank you for your interest in our project. We have discussed and we think it would be very nice and productive to schedule a call between us, in order to find common points and discuss possible ideas and solutions.
Please let us know.

Best regards,

Matteo
________________________________

**Comment by Bonifatius94** [comment]
Sure, we can have a call. Let me know which messenger you wanna use (e.g. Zoom, MS Teams or Discord). Maybe my supervisor at the University of Augsburg wants to join as well (@ll7).

For my part, I think it's really about performance. With zero training payload, the environment is currently capable of processing about 2000 steps per minute which equals roughly 30 training episodes (given lots of episodes fail right at the start, so it's probably a lot less once the robot learns how to move). From my experience with reinforcement learning, a training potentially requires millions of episodes to produce reasonable results. Applying some basic math, 1 million episodes requires 1,000,000 / 30 * 1 min = 33,333 minutes = 556 hours = 23 days. Even if you're highly parallelizing the training load with e.g. A3C, the training time is still not feasible. Maybe you can provide some insights on your past training experiments to put this totally made up calculation into perspective.

I don't know how much your cython build improved the simulation performance. But I'd say we need to push these 23 days to not more than a few hours, which I think is totally possible with a bit of cython/numba optimization effort. For example, the LiDAR sensor's raycast algorithm seems to consume about half of the simulation time. Another huge potential for improvement is releasing the Python runtime GIL and removing the Python C-API overhead by calling directly into NumPy DLLs, etc. Finally, there are compiler intrinsics / SIMD instructions which could really come in handy for some low-level computations like bitwise AND on 128/256/512 bit registers. There's really lots of potential with little to medium effort.

**Comment by matteocaruso1993** [comment]
For us it would be perfect either next Tuesday or next Wednesday. As for the platform I would suggest MS Teams.
Best,

Matteo
________________________________

**Comment by Bonifatius94** [comment]
Hey Matteo,
My supervisor wants to join as well. Tuesday would be fine with us (any time).
Best wishes, Marco

**Comment by Bonifatius94** [comment]
Hey guys, just a quick question. Would it be possible to constrain tomorrows meeting window to early afternoon?
I'd like to attend a funeral together with my mom. No problem if it's not possible.
Thanks in advance and see you tomorrow.
Best wishes, Marco

**Comment by matteocaruso1993** [comment]
Yes, sure it's fine for us in the early afternoon. Can you please send me your and your supervisor email such that I can invite you in the team for tomorrow?
Best,

Matteo
________________________________

**Comment by Bonifatius94** [comment]
Hey guys,

I finally did some pygame visualization of the simulator for visual debugging. Apparently, there seems to be an issue with the pedestrians controlled by ExtdSimulator. They seem to run through walls / obstacles. Are you aware of those issues? I guess we need to model some forces to keep the pedestrians away from obstacles, right? I hope this doesn't have any severe implications regarding your training results.

Additionally, I've been setting up lots of tools for profiling / collaboration such that it could be beneficial to do the PR now. I'm planning to upload my changes on my own GitHub repo soon. According to my benchmarks, the simulation performance is currently limited by PySocialForce itself, so I'll have to optimize that one as well. But yeah, shouldn't be too hard to pull off. I'm still kind of optimistic considering the past progress I made.

Best wishes,
Marco

**Comment by Bonifatius94** [comment]
I've been digging into PySocialForce lately and identified the ObstacleForce as the bottleneck, taking 98% of time. With the help of the PySocialForce repo owner, I figured out a much more efficient way to compute the force, so now the SocialForce dominates the benchmark.

Moreover, I've created a very basic OpenAI StableBaselines training with robot_sf. It's capable of training 800k steps/h using the A2C algorithm (on my private machine with RTX 3060 and Intel i5 6500, 4 cores, 16 GB RAM).

Do you think this performance improvement is good enough to make retraining your model feasible? :wink:
